integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingSynchronizationStrategyIT.java

```diff
@@ -0,0 +1,285 @@
+				 * until the work future is complete.
+				 */
+				() -> assertThat( workFuture ).isSuccessful()
+		);
+
+		// We expect the transaction to block forever, because the work future isn't complete
+		ALMOST_FOREVER_UNIT.sleep( ALMOST_FOREVER_VALUE );
+		assertThat( transactionFuture ).isPending();
+
+		// Completing the work should allow the transaction to unblock the thread
+		workFuture.complete( null );
+		/*
+		 * Note that this will throw an ExecutionException it the transaction failed
+		 * or an assertion failed in the other thread.
+		 */
+		transactionFuture.get( ALMOST_FOREVER_VALUE, ALMOST_FOREVER_UNIT );
+		assertThat( transactionFuture ).isSuccessful();
+	}
+
+	private void testAsynchronous(SessionFactory sessionFactory,
+			DocumentRefreshStrategy expectedRefreshStrategy)
+			throws InterruptedException, ExecutionException, TimeoutException {
+		CompletableFuture<?> workFuture = new CompletableFuture<>();
+
+		CompletableFuture<?> transactionFuture = runTransactionInDifferentThread(
+				sessionFactory,
+				null,
+				expectedRefreshStrategy,
+				workFuture,
+				/*
+				 * With this synchronization strategy, the transaction will unblock the thread
+				 * before the work future is complete.
+				 */
+				() -> assertThat( workFuture ).isPending()
+		);
+
+		/*
+		 * We didn't complete the work, but the transaction should unblock the thread anyway.
+		 * Note that this will throw an ExecutionException it the transaction failed
+		 * or an assertion failed in the other thread.
+		 */
+		transactionFuture.get( ALMOST_FOREVER_VALUE, ALMOST_FOREVER_UNIT );
+	}
+
+	/*
+	 * Run a transaction in a different thread so that its progress can be inspected from the current thread.
+	 */
+	private CompletableFuture<?> runTransactionInDifferentThread(SessionFactory sessionFactory,
+			AutomaticIndexingSynchronizationStrategy customStrategy,
+			DocumentRefreshStrategy expectedRefreshStrategy,
+			CompletableFuture<?> workFuture,
+			Runnable afterTransactionAssertion)
+			throws InterruptedException, ExecutionException, TimeoutException {
+		CompletableFuture<?> justBeforeTransactionCommitFuture = new CompletableFuture<>();
+		CompletableFuture<?> transactionFuture = CompletableFuture.runAsync( () -> {
+			OrmUtils.withinTransaction( sessionFactory, session -> {
+				if ( customStrategy != null ) {
+					Search.getSearchSession( session ).setAutomaticIndexingSynchronizationStrategy( customStrategy );
+				}
+				IndexedEntity entity1 = new IndexedEntity();
+				entity1.setId( 1 );
+				entity1.setIndexedField( "initialValue" );
+
+				session.persist( entity1 );
+
+				backendMock.expectWorks( IndexedEntity.INDEX, expectedRefreshStrategy )
+						.add( "1", b -> b
+								.field( "indexedField", entity1.getIndexedField() )
+						)
+						.preparedThenExecuted( workFuture );
+				justBeforeTransactionCommitFuture.complete( null );
+			} );
+			backendMock.verifyExpectationsMet();
+
+			afterTransactionAssertion.run();
```

Comment:
Yes, because we need to run assertions while the transaction commit is ongoing. We need to execute code in parallel of the transaction commit.
See `testSynchronous` in particular.